File: nova_lang/parser.py
"""
Recursive descent parser for NovaLang
"""

from .token import TokenType
from .ast_nodes import *

class ParserError(Exception):
    pass

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.i = 0
        self.current = tokens[0]

    def advance(self):
        self.i += 1
        if self.i < len(self.tokens):
            self.current = self.tokens[self.i]
        return self.current

    def match(self, *types):
        if self.current.type in types:
            t = self.current
            self.advance()
            return t
        expected = ', '.join(t.name for t in types)
        raise ParserError(f"Expected {expected} at line {self.current.line}, col {self.current.col}, got {self.current.type}")

    def parse(self):
        # Expect start ... end
        self.match(TokenType.START)
        statements = self.statements()
        self.match(TokenType.END)
        self.match(TokenType.EOF)
        return Program(statements)

    def statements(self):
        stmts = []
        while self.current.type not in (TokenType.END, TokenType.RBRACE, TokenType.EOF):
            stmts.append(self.statement())
        return stmts

    def statement(self):
        t = self.current.type
        if t in (TokenType.NUM, TokenType.TEXT, TokenType.FLAG):
            return self.var_decl()
        elif t == TokenType.IDENT:
            return self.assign_or_func_call()
        elif t == TokenType.SHOW:
            return self.show_stmt()
        elif t == TokenType.TAKE:
            return self.take_stmt()
        elif t == TokenType.WHEN:
            return self.when_stmt()
        elif t == TokenType.LOOP:
            return self.loop_stmt()
        elif t == TokenType.BREAK:
            self.match(TokenType.BREAK)
            return Break()
        elif t == TokenType.FUNC:
            return self.func_def()
        else:
            raise ParserError(f"Unexpected token {self.current.type} at line {self.current.line}")

    def var_decl(self):
        vartype_token = self.match(TokenType.NUM, TokenType.TEXT, TokenType.FLAG)
        vartype = vartype_token.value
        name = self.match(TokenType.IDENT).value
        self.match(TokenType.ASSIGN)
        expr = self.expr()
        return VarDecl(vartype, name, expr)

    def assign_or_func_call(self):
        name = self.match(TokenType.IDENT).value
        if self.current.type == TokenType.ASSIGN:
            self.match(TokenType.ASSIGN)
            expr = self.expr()
            return Assign(name, expr)
        elif self.current.type == TokenType.LPAREN:
            # func call
            self.match(TokenType.LPAREN)
            args = []
            if self.current.type != TokenType.RPAREN:
                args.append(self.expr())
                while self.current.type == TokenType.COMMA:
                    self.match(TokenType.COMMA)
                    args.append(self.expr())
            self.match(TokenType.RPAREN)
            return FuncCall(name, args)
        else:
            raise ParserError(f"Expected assignment or func call at line {self.current.line}")

    def show_stmt(self):
        self.match(TokenType.SHOW)
        expr = self.expr()
        return Show(expr)

    def take_stmt(self):
        self.match(TokenType.TAKE)
        name = self.match(TokenType.IDENT).value
        return Take(name)

    def when_stmt(self):
        cases = []
        else_block = None

        # when cond { stmts }
        self.match(TokenType.WHEN)
        cond = self.expr()
        block = self.block()
        cases.append((cond, block))

        # elsewhen*
        while self.current.type == TokenType.ELSEWHEN:
            self.match(TokenType.ELSEWHEN)
            cond = self.expr()
            block = self.block()
            cases.append((cond, block))

        # else optional
        if self.current.type == TokenType.ELSE:
            self.match(TokenType.ELSE)
            else_block = self.block()

        return When(cases, else_block)

    def loop_stmt(self):
        self.match(TokenType.LOOP)
        var = self.match(TokenType.IDENT).value
        self.match(TokenType.ASSIGN)
        start_expr = self.expr()
        self.match(TokenType.TO)
        end_expr = self.expr()
        body = self.block()
        return Loop(var, start_expr, end_expr, body)

    def func_def(self):
        self.match(TokenType.FUNC)
        name = self.match(TokenType.IDENT).value
        self.match(TokenType.LPAREN)
        params = []
        if self.current.type != TokenType.RPAREN:
            params.append(self.match(TokenType.IDENT).value)
            while self.current.type == TokenType.COMMA:
                self.match(TokenType.COMMA)
                params.append(self.match(TokenType.IDENT).value)
        self.match(TokenType.RPAREN)
        # body { stmts back expr }
        self.match(TokenType.LBRACE)
        body = []
        back_expr = None
        while self.current.type != TokenType.BACK:
            if self.current.type == TokenType.RBRACE:
                raise ParserError("Function must contain a 'back' statement")
            body.append(self.statement())
        self.match(TokenType.BACK)
        back_expr = self.expr()
        self.match(TokenType.RBRACE)
        return FuncDef(name, params, body, back_expr)

    def block(self):
        self.match(TokenType.LBRACE)
        stmts = self.statements()
        self.match(TokenType.RBRACE)
        return stmts

    # Expression grammar:
    # expr -> equality
    def expr(self):
        return self.equality()

    def equality(self):
        node = self.comparison()
        while self.current.type in (TokenType.EQEQ, TokenType.NOTEQ):
            op = self.current
            self.advance()
            right = self.comparison()
            node = BinOp(node, op, right)
        return node

    def comparison(self):
        node = self.term()
        while self.current.type in (TokenType.GT, TokenType.LT, TokenType.GTEQ, TokenType.LTEQ):
            op = self.current
            self.advance()
            right = self.term()
            node = BinOp(node, op, right)
        return node

    def term(self):
        node = self.factor()
        while self.current.type in (TokenType.PLUS, TokenType.MINUS):
            op = self.current
            self.advance()
            right = self.factor()
            node = BinOp(node, op, right)
        return node

    def factor(self):
        node = self.unary()
        while self.current.type in (TokenType.STAR, TokenType.SLASH):
            op = self.current
            self.advance()
            right = self.unary()
            node = BinOp(node, op, right)
        return node

    def unary(self):
        if self.current.type in (TokenType.MINUS,):
            op = self.current
            self.advance()
            expr = self.unary()
            return UnaryOp(op, expr)
        return self.primary()

    def primary(self):
        t = self.current
        if t.type == TokenType.NUMBER:
            self.advance()
            return Literal(int(t.value), 'num')
        elif t.type == TokenType.STRING:
            self.advance()
            return Literal(t.value.strip('"'), 'text')
        elif t.type == TokenType.BOOL:
            self.advance()
            return Literal(True if t.value == 'true' else False, 'bool')
        elif t.type == TokenType.IDENT:
            # Could be func call
            name = t.value
            self.advance()
            if self.current.type == TokenType.LPAREN:
                # call
                self.match(TokenType.LPAREN)
                args = []
                if self.current.type != TokenType.RPAREN:
                    args.append(self.expr())
                    while self.current.type == TokenType.COMMA:
                        self.match(TokenType.COMMA)
                        args.append(self.expr())
                self.match(TokenType.RPAREN)
                return FuncCall(name, args)
            return Identifier(name)
        elif t.type == TokenType.LPAREN:
            self.match(TokenType.LPAREN)
            node = self.expr()
            self.match(TokenType.RPAREN)
            return node
        else:
            raise ParserError(f"Unexpected token {t.type} at line {t.line}")

File: nova_lang/Parser.cpp
// Parser.cpp — corrected version

#include "parser.hpp"
#include <sstream>
#include <stdexcept>

Parser::Parser(const std::vector<Token>& toks) : tokens(toks), i(0) {
    if (tokens.empty()) throw ParserError("Empty token stream");
}

const Token& Parser::current() const {
    if (i < tokens.size()) return tokens[i];
    static Token eof_t(TokenType::EOF_T, "", 0, 0);
    return eof_t;
}

Token Parser::advance() {
    if (i < tokens.size()) return tokens[i++];
    return Token(TokenType::EOF_T, "", tokens.back().line, tokens.back().col);
}

Token Parser::match(std::initializer_list<TokenType> types) {
    for (auto t : types) {
        if (current().type == t) {
            return advance();
        }
    }
    std::ostringstream ss;
    ss << "Expected token(s) at " << current().line << ":" << current().col;
    throw ParserError(ss.str());
}

std::unique_ptr<Program> Parser::parse() {
    match({TokenType::START});
    StmtList stmts = statements();
    match({TokenType::END});
    match({TokenType::EOF_T});
    return std::make_unique<Program>(std::move(stmts));
}

StmtList Parser::statements() {
    StmtList stmts;
    while (current().type != TokenType::END &&
           current().type != TokenType::RBRACE &&
           current().type != TokenType::EOF_T) {
        stmts.push_back(statement());
    }
    return stmts;
}

ASTNodePtr Parser::statement() {
    TokenType t = current().type;
    if (t == TokenType::NUM || t == TokenType::TEXT || t == TokenType::FLAG) return var_decl();
    else if (t == TokenType::IDENT) return assign_or_func_call();
    else if (t == TokenType::SHOW) return show_stmt();
    else if (t == TokenType::TAKE) return take_stmt();
    else if (t == TokenType::WHEN) return when_stmt();
    else if (t == TokenType::LOOP) return loop_stmt();
    else if (t == TokenType::BREAK) { match({TokenType::BREAK}); return std::make_unique<Break>(); }
    else if (t == TokenType::FUNC) return func_def();
    else {
        std::ostringstream ss; ss << "Unexpected token at " << current().line << ":" << current().col; 
        throw ParserError(ss.str());
    }
}

ASTNodePtr Parser::var_decl() {
    Token vt = match({TokenType::NUM, TokenType::TEXT, TokenType::FLAG});
    Token name = match({TokenType::IDENT});
    match({TokenType::ASSIGN});
    ASTNodePtr ex = expr();
    return std::make_unique<VarDecl>(vt.value, name.value, std::move(ex));
}

ASTNodePtr Parser::assign_or_func_call() {
    Token name = match({TokenType::IDENT});
    if (current().type == TokenType::ASSIGN) {
        match({TokenType::ASSIGN});
        ASTNodePtr e = expr();
        return std::make_unique<Assign>(name.value, std::move(e));
    } else if (current().type == TokenType::LPAREN) {
        match({TokenType::LPAREN});
        std::vector<ASTNodePtr> args;
        if (current().type != TokenType::RPAREN) {
            args.push_back(expr());
            while (current().type == TokenType::COMMA) { match({TokenType::COMMA}); args.push_back(expr()); }
        }
        match({TokenType::RPAREN});
        return std::make_unique<FuncCall>(name.value, std::move(args));
    } else {
        std::ostringstream ss; ss << "Expected assign or func-call at " << current().line << ":" << current().col; throw ParserError(ss.str());
    }
}

ASTNodePtr Parser::show_stmt() { match({TokenType::SHOW}); ASTNodePtr e = expr(); return std::make_unique<Show>(std::move(e)); }
ASTNodePtr Parser::take_stmt() { match({TokenType::TAKE}); Token id = match({TokenType::IDENT}); return std::make_unique<Take>(id.value); }

ASTNodePtr Parser::when_stmt() {
    // We'll collect cases as (condition ASTNodePtr, StmtList)
    std::vector<std::pair<ASTNodePtr, StmtList>> cases;
    StmtList else_block;

    Token when_tok = match({TokenType::WHEN});
    ASTNodePtr cond = expr();

    // block() returns ASTNodePtr (Program). Extract statements from it.
    ASTNodePtr block_node = block(when_tok);
    Program* prog = dynamic_cast<Program*>(block_node.get());
    if (!prog) throw ParserError("Internal parser error: expected Program node from block()");
    StmtList body = std::move(prog->statements);
    cases.emplace_back(std::move(cond), std::move(body));

    while (current().type == TokenType::ELSEWHEN) {
        Token t = match({TokenType::ELSEWHEN});
        ASTNodePtr cond2 = expr();
        ASTNodePtr block_node2 = block(t);
        Program* prog2 = dynamic_cast<Program*>(block_node2.get());
        if (!prog2) throw ParserError("Internal parser error: expected Program node from block()");
        StmtList b2 = std::move(prog2->statements);
        cases.emplace_back(std::move(cond2), std::move(b2));
    }

    if (current().type == TokenType::ELSE) {
        Token t = match({TokenType::ELSE});
        ASTNodePtr else_node = block(t);
        Program* prog_else = dynamic_cast<Program*>(else_node.get());
        if (!prog_else) throw ParserError("Internal parser error: expected Program node from block()");
        else_block = std::move(prog_else->statements);
    }

    return std::make_unique<When>(std::move(cases), std::move(else_block));
}

ASTNodePtr Parser::loop_stmt() {
    Token loop_t = match({TokenType::LOOP});
    Token var = match({TokenType::IDENT});
    match({TokenType::ASSIGN});
    ASTNodePtr s = expr();
    match({TokenType::TO});
    ASTNodePtr e = expr();

    ASTNodePtr block_node = block(loop_t);
    Program* prog = dynamic_cast<Program*>(block_node.get());
    if (!prog) throw ParserError("Internal parser error: expected Program node from block()");
    StmtList body = std::move(prog->statements);

    return std::make_unique<Loop>(var.value, std::move(s), std::move(e), std::move(body));
}

ASTNodePtr Parser::func_def() {
    match({TokenType::FUNC});
    Token name = match({TokenType::IDENT});
    match({TokenType::LPAREN});
    std::vector<std::string> params;
    if (current().type != TokenType::RPAREN) {
        params.push_back(match({TokenType::IDENT}).value);
        while (current().type == TokenType::COMMA) { match({TokenType::COMMA}); params.push_back(match({TokenType::IDENT}).value); }
    }
    match({TokenType::RPAREN});
    match({TokenType::LBRACE});
    StmtList body;
    ASTNodePtr back_expr;
    while (current().type != TokenType::BACK) {
        if (current().type == TokenType::RBRACE) throw ParserError("Function must contain a 'back' statement");
        body.push_back(statement());
    }
    match({TokenType::BACK});
    back_expr = expr();
    match({TokenType::RBRACE});
    return std::make_unique<FuncDef>(name.value, std::move(params), std::move(body), std::move(back_expr));
}

ASTNodePtr Parser::block(const Token& context_token) {
    if (current().type != TokenType::LBRACE) {
        std::ostringstream ss;
        ss << "Expected LBRACE at " << context_token.line << ":" << context_token.col << " (after '" << context_token.value << "')";
        throw ParserError(ss.str());
    }
    match({TokenType::LBRACE});
    StmtList stmts = statements();
    match({TokenType::RBRACE});
    // Return a Program node containing the block statements
    return std::make_unique<Program>(std::move(stmts));
}

// Expressions
ASTNodePtr Parser::expr() { return equality(); }

ASTNodePtr Parser::equality() {
    ASTNodePtr node = comparison();
    while (current().type == TokenType::EQEQ || current().type == TokenType::NOTEQ) {
        Token op = advance();
        ASTNodePtr right = comparison();
        node = std::make_unique<BinOp>(std::move(node), op.type, op.value, std::move(right));
    }
    return node;
}

ASTNodePtr Parser::comparison() {
    ASTNodePtr node = term();
    while (current().type == TokenType::GT || current().type == TokenType::LT || current().type == TokenType::GTEQ || current().type == TokenType::LTEQ) {
        Token op = advance();
        ASTNodePtr right = term();
        node = std::make_unique<BinOp>(std::move(node), op.type, op.value, std::move(right));
    }
    return node;
}

ASTNodePtr Parser::term() {
    ASTNodePtr node = factor();
    while (current().type == TokenType::PLUS || current().type == TokenType::MINUS) {
        Token op = advance();
        ASTNodePtr right = factor();
        node = std::make_unique<BinOp>(std::move(node), op.type, op.value, std::move(right));
    }
    return node;
}

ASTNodePtr Parser::factor() {
    ASTNodePtr node = unary();
    while (current().type == TokenType::STAR || current().type == TokenType::SLASH) {
        Token op = advance();
        ASTNodePtr right = unary();
        node = std::make_unique<BinOp>(std::move(node), op.type, op.value, std::move(right));
    }
    return node;
}

ASTNodePtr Parser::unary() {
    if (current().type == TokenType::MINUS) {
        Token op = advance();
        ASTNodePtr e = unary();
        return std::make_unique<UnaryOp>(op.type, op.value, std::move(e));
    }
    return primary();
}

ASTNodePtr Parser::primary() {
    Token t = current();
    if (t.type == TokenType::NUMBER) {
        advance();
        return std::make_unique<Literal>(t.value, "num");
    } else if (t.type == TokenType::STRING) {
        advance();
        return std::make_unique<Literal>(t.value, "text");
    } else if (t.type == TokenType::BOOL) {
        advance();
        return std::make_unique<Literal>(t.value, "bool");
    } else if (t.type == TokenType::IDENT) {
        advance();
        if (current().type == TokenType::LPAREN) {
            match({TokenType::LPAREN});
            std::vector<ASTNodePtr> args;
            if (current().type != TokenType::RPAREN) {
                args.push_back(expr());
                while (current().type == TokenType::COMMA) { match({TokenType::COMMA}); args.push_back(expr()); }
            }
            match({TokenType::RPAREN});
            return std::make_unique<FuncCall>(t.value, std::move(args));
        }
        return std::make_unique<Identifier>(t.value);
    } else if (t.type == TokenType::LPAREN) {
        match({TokenType::LPAREN});
        ASTNodePtr n = expr();
        match({TokenType::RPAREN});
        return n;
    }
    std::ostringstream ss; ss << "Unexpected token in expression at " << t.line << ":" << t.col; throw ParserError(ss.str());
}
